src/patch.zig

```diff
@@ -0,0 +1,1483 @@
+                    .path = destination_file_path.?,
+                    .hunks = file.takeHunks(),
+                    .before_hash = file.before_hash,
+                    .after_hash = file.after_hash,
+                }),
+            }) catch unreachable;
+        }
+    }
+
+    return result;
+}
+
+fn parseFileMode(mode: []const u8) ?FileMode {
+    const parsed_mode = (std.fmt.parseInt(u32, mode, 8) catch return null) & 0o777;
+    return FileMode.fromU32(parsed_mode);
+}
+
+const LookbackIterator = struct {
+    inner: std.mem.SplitIterator(u8, .scalar),
+    prev_index: usize = 0,
+
+    pub fn fromInner(inner: std.mem.SplitIterator(u8, .scalar)) LookbackIterator {
+        return LookbackIterator{ .inner = inner };
+    }
+
+    pub fn next(this: *LookbackIterator) ?[]const u8 {
+        this.prev_index = this.inner.index orelse this.prev_index;
+        return this.inner.next();
+    }
+
+    pub fn back(this: *LookbackIterator) void {
+        this.inner.index = this.prev_index;
+    }
+};
+
+const PatchLinesParser = struct {
+    result: List(FileDeets) = .{},
+    current_file_patch: FileDeets = .{},
+    state: State = .parsing_header,
+    current_hunk: ?Hunk = null,
+    current_hunk_mutation_part: ?PatchMutationPart = null,
+
+    const State = enum { parsing_header, parsing_hunks };
+
+    const HunkLineType = enum(u3) {
+        /// Additional context
+        context = 0,
+
+        /// Example:
+        /// + sjfskdjfsdf
+        insertion,
+
+        /// Example:
+        /// - sjfskdjfsdf
+        deletion,
+
+        /// Example:
+        /// @@ -1,3 +1,3 @@
+        header,
+
+        /// Example:
+        /// \ No newline at end of file
+        pragma,
+    };
+
+    fn deinit(this: *PatchLinesParser, allocator: Allocator, comptime clear_result_retaining_capacity: bool) void {
+        this.current_file_patch.deinit(allocator);
+        if (this.current_hunk) |*hunk| hunk.deinit(allocator);
+        if (this.current_hunk_mutation_part) |*part| part.deinit(allocator);
+        for (this.result.items) |*file_deet| file_deet.deinit(allocator);
+        if (comptime clear_result_retaining_capacity) {
+            this.result.clearRetainingCapacity();
+        } else {
+            this.result.deinit(allocator);
+        }
+    }
+
+    fn reset(this: *PatchLinesParser, allocator: Allocator) void {
+        this.deinit(allocator, true);
+        this.result.clearRetainingCapacity();
+        this.* = .{
+            .result = this.result,
+        };
+    }
+
+    pub fn parse(
+        this: *PatchLinesParser,
+        file_: []const u8,
+        opts: struct { support_legacy_diffs: bool = false },
+    ) ParseErr!void {
+        const end = brk: {
+            var iter = std.mem.splitBackwardsScalar(u8, file_, '\n');
+            var prev: usize = file_.len;
+            if (iter.next()) |last_line| {
+                if (last_line.len == 0) {
+                    prev = iter.index.?;
```

Comment:
```suggestion
                    prev = iter.index orelse return;
```
Hit this with an empty patch file. Maybe a check for len == 0 earlier would be better